pkg/analyze/event.go

```diff
@@ -0,0 +1,218 @@
+	reason   string
+	msgRegex string
+}
+
+func (a *AnalyzeEvent) Title() string {
+	if a.analyzer.CheckName != "" {
+		return a.analyzer.CheckName
+	}
+	if a.analyzer.CollectorName != "" {
+		return a.analyzer.CollectorName
+	}
+	return "Event"
+}
+
+func (a *AnalyzeEvent) IsExcluded() (bool, error) {
+	return isExcluded(a.analyzer.Exclude)
+}
+
+func (a *AnalyzeEvent) Analyze(getFile getCollectedFileContents, findFiles getChildCollectedFileContents) ([]*AnalyzeResult, error) {
+	// required check
+	if a.analyzer.Reason == "" {
+		return nil, errors.New("reason is required")
+	}
+
+	// read collected events based on namespace
+	namespace := getNamespace(a.analyzer.Namespace)
+	fullPath := path.Join(constants.CLUSTER_RESOURCES_DIR, constants.CLUSTER_RESOURCES_EVENTS, namespace)
+	fullPath = fmt.Sprintf("%s.json", fullPath)
+	fileContent, err := getFile(fullPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read collected events for namespace: %s", namespace)
+	}
+
+	events, err := convertToEventList(fileContent)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read collected events")
+	}
+
+	// filter if there's single event matched with the given criteria
+	// match: Reason && Kind (optional) && MessageRegex (optional)
+	// e.g. Reason: Unhealthy. Kind: Pod. Message: Readiness probe failed:...
+	event := getEvent(events, eventFilter{
+		kind:     a.analyzer.Kind,
+		reason:   a.analyzer.Reason,
+		msgRegex: a.analyzer.RegexPattern,
+	})
+
+	return analyzeEventResult(event, a.analyzer.Outcomes, a.Title())
+
+}
+
+func getNamespace(namespace string) string {
+	if namespace == "" {
+		return corev1.NamespaceDefault
+	}
+	return namespace
+}
+
+func convertToEventList(data []byte) (*corev1.EventList, error) {
+	var eventList corev1.EventList
+	err := json.Unmarshal(data, &eventList)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert []byte to corev1.EventList: %w", err)
+	}
+	return &eventList, nil
+}
+
+func getEvent(events *corev1.EventList, filter eventFilter) *corev1.Event {
+	for _, event := range events.Items {
+		if !matchReason(event.Reason, filter.reason) {
+			continue
+		}
+		if !matchKind(event.InvolvedObject.Kind, filter.kind) {
+			continue
+		}
+		if matchMessage(event.Message, filter.msgRegex) {
+			klog.V(2).Infof("event matched: %v for reason: %s kind: %s messageRegex: %s ", event, filter.reason, filter.kind, filter.msgRegex)
+			return &event
+		}
+	}
+	return nil
+}
+
+func matchReason(actual, expected string) bool {
+	// not possible to have empty reason
+	if expected == "" {
+		return false
+	}
+	return strings.EqualFold(actual, expected)
+}
+
+func matchKind(actual, expected string) bool {
+	// kind is optional
+	if expected == "" {
+		return true
+	}
+	return strings.EqualFold(actual, expected)
+}
+
+func matchMessage(actual, expectedRegex string) bool {
+	// message Regex is optional
+	if expectedRegex == "" {
+		return true
+	}
+	re, err := regexp.Compile(expectedRegex)
```

Comment:
Minor: I'd compile this once before iterating over the events

Reply:
many thanks @banjoh! that's a great catch! I have updated the code.